### PR TITLE
Fixes for shellcheck warnings, part 1

### PR DESCRIPTION
**Pull request by cameel**
~Depends on #10877.~ It's on `develop` now.

This PR fixes most of shellcheck warnings in our scripts. Those are only the ones that were not actual errors (those are covered by #10584, #10583) or just missing quotes (#10585, #10877).

~These are not all fixes, just the ones that I could easily put in an independent branch. I have a few more but they depend on all the other branches so I'd like to get these here merged first. When we're done with them I'll submit the rest and another PR that removes all the scripts (now warning-free) from shellcheck ignore list.~ ~EDIT: I decided to just put it on top of the other PRs and just get them all through review one by one. So this will remain a draft until all the PRs below are merged.~ EDIT: Ready for review.

### CODE REVIEW

File: .circleci/soltest_all.sh [hunk 1]
```diff
@@ -32,38 +32,35 @@ EVM_VALUES=(homestead byzantium constantinople petersburg istanbul)
...
+        RUN_STEPS=$(seq 2 "$STEPS" | circleci tests split | xargs)
     fi
 else
-    RUN_STEPS=$(seq "$STEPS")
+    RUN_STEPS=$(seq "$STEPS" | xargs)
 fi
 
-# turn newlines into spaces
-RUN_STEPS=$(echo $RUN_STEPS)
-
```
Comment (cameel) [inline .circleci/soltest_all.sh:55]: So we ended up with https://github.com/ethereum/solidity/pull/9421#discussion_r455112574 anyway...

File: scripts/wasm-rebuild/docker-scripts/rebuild_current.sh [hunk 2]
```diff
@@ -41,7 +41,14 @@ if [ -d jsoncpp ]; then
 fi
 
 set +e
-scripts/*/build_emscripten.sh
+
+if [ -e scripts/c1/build_emscripten.sh ]; then
```
Comment (axic) [inline scripts/wasm-rebuild/docker-scripts/rebuild_current.sh]:
```suggestion
if [ -e scripts/ci/build_emscripten.sh ]; then
```

Comment (cameel) [reply scripts/wasm-rebuild/docker-scripts/rebuild_current.sh]: Thanks. I thought I fixed it but turns out that only on the line below :(

**Comment by leonardoalt** [comment]
@cameel why is it a draft?

File: scripts/check_style.sh [hunk 3]
```diff
@@ -24,17 +24,17 @@ function preparedGrep()
...
 	preparedGrep "[,\(<]\s*const " # const on left side of type
 	preparedGrep "^\s*(static)?\s*const " # const on left side of type (beginning of line)
 	preparedGrep "^ [^*]|[^*] 	|	 [^*]" # uses spaces for indentation or mixes spaces and tabs
-	preparedGrep "[a-zA-Z0-9_]\s*[&][a-zA-Z_]" | egrep -v "return [&]" # right-aligned reference ampersand (needs to exclude return)
+	preparedGrep "[a-zA-Z0-9_]\s*[&][a-zA-Z_]" | grep -E -v "return [&]" # right-aligned reference ampersand (needs to exclude return)
 	# right-aligned reference pointer star (needs to exclude return and comments)
-	preparedGrep "[a-zA-Z0-9_]\s*[*][a-zA-Z_]" | egrep -v -e "return [*]" -e "^* [*]" -e "^*//.*"
-) | egrep -v -e "^[a-zA-Z\./]*:[0-9]*:\s*\/(\/|\*)" -e "^test/"
+	preparedGrep "[a-zA-Z0-9_]\s*[*][a-zA-Z_]" | grep -E -v -e "return [*]" -e "^* [*]" -e "^*//.*"
+) | grep -E -v -e "^[a-zA-Z\./]*:[0-9]*:\s*\/(\/|\*)" -e "^test/"
```
Comment (leonardoalt) [inline scripts/check_style.sh]: Does this work on MacOS as well? I think we had some problems in the past because of `grep -E`, but might be imagining that

File: scripts/endToEndExtraction/create_traces.sh [hunk 4]
```diff
@@ -1,22 +1,25 @@
-BASE_PATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 || exit ; pwd -P )"
+#!/usr/bin/env bash
+set -eu
+
+BASE_PATH="$( cd "$(dirname "$0")" >/dev/null 2>&1 && pwd -P )"
 
 mkdir -p build
-cd build || exit
```
Comment (leonardoalt) [inline scripts/endToEndExtraction/create_traces.sh:4]: Why were these removed?

Comment (cameel) [reply scripts/endToEndExtraction/create_traces.sh:4]: Because I added `set -eu` at the top of the script so it will fail automatically if `cd` fails.

File: scripts/check_style.sh (hunk 3 above)
Comment (cameel) [reply scripts/check_style.sh]: I haven't checked that yet. This is the reason why this and #10585 are still drafts - I need to test this stuff a bit more because we have very little coverage for shell scripts but I want to finish the bytecode compare PRs first.

**Comment by cameel** [comment]
> @cameel why is it a draft?

Basically, I'm not yet 100% confident that this does not break anything and I want to test it manually a bit more but I'm stuck in the bytecode compare PRs at the moment and I want to finish them first.

**Comment by axic** [comment]
I think you could definitely merge without hesitation the `/usr/bin/env` and `set -e` changes separately.

File: scripts/ASTImportTest.sh [hunk 5]
```diff
@@ -58,9 +58,9 @@ function testImportExportEquivalence {
             then
                 echo -e "ERROR: JSONS differ for $1: \n $DIFF \n"
                 echo "Expected:"
-                echo "$(cat ./expected.json)"
+                cat ./expected.json
                 echo "Obtained:"
-                echo "$(cat ./obtained.json)"
+                cat ./obtained.json
```
Comment (axic) [inline scripts/ASTImportTest.sh:65]: Also this is pretty clearly not breaking anything?

File: scripts/check_style.sh [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 (
 REPO_ROOT="$(dirname "$0")"/..
-cd $REPO_ROOT
+cd $REPO_ROOT || exit 1
```
Comment (axic) [inline scripts/check_style.sh]: Why not just `set -e`?

Comment (cameel) [reply scripts/check_style.sh]: Same as in the other PR. I didn't want to do it in such a big batch of changes to avoid introducing bugs.

**Comment by cameel** [comment]
@axic Yeah, that sounds like a good idea. I'm going to split this PR into smaller ones and get it merged in smaller pieces to make it more manageable. The changes you mentioned will go in first.

**Comment by cameel** [comment]
I tested all I could locally and I think this should all work fine. I'm marking it as ready for review.

File: scripts/check_style.sh (hunk 3 above)
Comment (cameel) [reply scripts/check_style.sh]: This should be fine on macOS. `-E` is one of the options that work the same way between Linux and macOS. AFAIK `-P` (`--perl-regexp`) that we're using in `get_version.sh` is the problematic one.

File: scripts/check_style.sh (hunk 6 above)
Comment (cameel) [reply scripts/check_style.sh]: Done. This required ignoring the return values from `grep`.

File: scripts/wasm-rebuild/docker-scripts/genbytecode.sh [hunk 7]
```diff
@@ -93,6 +93,6 @@ for (var optimize of [false, true])
 }
 EOF
     chmod +x solc
-    ./solc *.sol > /tmp/report.txt
+    ./solc -- *.sol > /tmp/report.txt
```
Comment (leonardoalt) [inline scripts/wasm-rebuild/docker-scripts/genbytecode.sh:96]: Why --?

File: scripts/wasm-rebuild/docker-scripts/rebuild_current.sh [hunk 8]
```diff
@@ -41,7 +41,14 @@ if [ -d jsoncpp ]; then
 fi
 
 set +e
-scripts/*/build_emscripten.sh
+
+if [ -e scripts/ci/build_emscripten.sh ]; then
+  scripts/ci/build_emscripten.sh
+else
+  # The script used to be in scripts/ci/ In earlier versions.
```
Comment (leonardoalt) [inline scripts/wasm-rebuild/docker-scripts/rebuild_current.sh]:
```suggestion
  # The script used to be in scripts/ci/ in earlier versions.
```

File: scripts/wasm-rebuild/docker-scripts/genbytecode.sh (hunk 7 above)
Comment (cameel) [reply scripts/wasm-rebuild/docker-scripts/genbytecode.sh:96]: If you have a file that's named like a command line option, `*` will match it. For example if there's a file called `-rf` and you do `rm *`, you're actually doing `rm -rf *` :) To avoid this the convention in many programs is that stuff after `--` is never interpreted as an option. So `rm -- *` is safe even if you have a file called `-rf`.

Another way to avoid that is to use `./*` instead of `*` because for most programs the difference between `x.txt` and `./x.txt` does not matter. Unfortunately for `solc` it does (though it might change soon: #4702). I used `./*` in most places that shellcheck pointed out becuase that's more universal but I could not do it here. Fortunately `boost::program_options` does support `--` so that's the workaround I used.

File: scripts/wasm-rebuild/docker-scripts/rebuild_current.sh (hunk 8 above)
Comment (cameel) [reply scripts/wasm-rebuild/docker-scripts/rebuild_current.sh]: Done.

File: scripts/wasm-rebuild/docker-scripts/genbytecode.sh (hunk 7 above)
Comment (leonardoalt) [reply scripts/wasm-rebuild/docker-scripts/genbytecode.sh:96]: Ah ok!

**Comment by leonardoalt** [comment]
Looks good, can you squash? I'll approve then

**Comment by cameel** [comment]
Why squash though? It's not a pile of random fixup commits :) Each commit deals with a different warning and the description says what is being fixed. It's just that there were quite a few different things to fix. It would be a jumbled mess if I squashed that so I'd rather not.

**Comment by leonardoalt** [comment]
Hmmm ok, it's quite a lot of commits though. I'll approve it anyway, if anyone doesn't like it blame @cameel !

**Review by leonardoalt — APPROVED** [review]

**Comment by cameel** [comment]
Yeah, I know. But do we have a rule against that? No one is being paid by the commit anyway :) I thought that it's more about whether it appropriate for the situation.

I do see that there's a tendency towards bigger commits in Solidity so I mostly try to do that too and only have a few per PR even though personally I prefer smaller ones. Especially in PRs like this one where the fix consists of many independent parts that are better off viewed separately, I think smaller ones make more sense. There are bigger and smaller changes here and I think that without separating them the smaller ones can easily get lost in the noise.

**Comment by leonardoalt** [comment]
That's fine with me
